// node-iframe/dist/iframe.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
exports.__esModule = true;
exports.fetchFrame = exports.configureTemplates = exports.configureResourceControl = exports.configureAgent = exports.appSourceConfig = void 0;
var fetch_1 = require("./fetch");
var cheerio_1 = require("cheerio");
var templates_1 = require("./templates");
exports.configureTemplates = templates_1.configureTemplates;
var config_1 = require("./config");
var appSourceConfig = config_1.defaultConfig;
exports.appSourceConfig = appSourceConfig;
var httpAgent;
var httpsAgent;
var agentConfigured = false;
var agent = agentConfigured
    ? function (_parsedURL) {
        if (_parsedURL.protocol == "http:" && setAgent(true)) {
            return httpAgent;
        }
        else if (setAgent(false)) {
            return httpsAgent;
        }
    }
    : undefined;
function setAgent(http) {
    try {
        if (http && !httpAgent) {
            var transport = require("http");
            httpAgent = new transport.Agent({
                rejectUnauthorized: false
            });
        }
        else if (!httpsAgent) {
            var transport = require("https");
            httpsAgent = new transport.Agent({
                rejectUnauthorized: false
            });
        }
        return true;
    }
    catch (e) {
        console.error(e);
        return false;
    }
}
function configureAgent() {
    if (!agentConfigured) {
        agentConfigured = true;
    }
}
exports.configureAgent = configureAgent;
var mutateSource = function (_a, url, $html, headers) {
    var _b = _a.src, src = _b === void 0 ? "" : _b, key = _a.key, safeMode = _a.safeMode;
    return __awaiter(void 0, void 0, void 0, function () {
        var res, source, _c, e_1;
        return __generator(this, function (_d) {
            switch (_d.label) {
                case 0:
                    if (!(src && src[0] === "/")) return [3, 8];
                    _d.label = 1;
                case 1:
                    _d.trys.push([1, 7, , 8]);
                    return [4, (0, fetch_1.fetcher)("".concat(url, "/").concat(src), {
                            headers: headers,
                            agent: agent
                        })];
                case 2:
                    res = _d.sent();
                    if (!res) return [3, 6];
                    if (!(typeof process !== "undefined")) return [3, 3];
                    _c = res;
                    return [3, 5];
                case 3: return [4, res.text()];
                case 4:
                    _c = _d.sent();
                    _d.label = 5;
                case 5:
                    source = _c;
                    $html(key).text(safeMode ? "try {".concat(source, "}catch(e){console.error(e)}") : source);
                    _d.label = 6;
                case 6: return [3, 8];
                case 7:
                    e_1 = _d.sent();
                    console.error(e_1);
                    return [3, 8];
                case 8: return [2];
            }
        });
    });
};
function renderErrorHtml(_a) {
    var url = _a.url, server = _a.server, _b = _a.noPage, noPage = _b === void 0 ? false : _b;
    return Object.assign((0, cheerio_1.load)(!url
        ? templates_1.templateModel[templates_1.TemplateType.error]
        : templates_1.templateModel[templates_1.TemplateType.notFound]), server ? { status: Number("".concat(40).concat(!url || noPage ? 4 : 0)) } : {});
}
function renderHtml(_a, server) {
    var url = _a.url, baseHref = _a.baseHref, config = _a.config, _b = _a.head, head = _b === void 0 ? {} : _b, safeMode = _a.safeMode;
    if (server === void 0) { server = false; }
    return __awaiter(this, void 0, void 0, function () {
        var _c, inline, cors, headers, response, e_2, html, _d, $html, inlineMutations_2, _loop_1, _i, _e, key, _f, inlineMutations_1, com, key, attribute, src, element, _g, _h, key, e_3;
        return __generator(this, function (_j) {
            switch (_j.label) {
                case 0:
                    if (!url) {
                        return [2, renderErrorHtml({ url: url, server: server })];
                    }
                    _c = {
                        inline: __assign(__assign({}, appSourceConfig.inline), config === null || config === void 0 ? void 0 : config.inline),
                        cors: __assign(__assign({}, appSourceConfig.cors), config === null || config === void 0 ? void 0 : config.cors)
                    }, inline = _c.inline, cors = _c.cors;
                    headers = __assign(__assign({}, config_1.appHeaders), head);
                    _j.label = 1;
                case 1:
                    _j.trys.push([1, 3, , 4]);
                    return [4, (0, fetch_1.fetcher)(url, {
                            headers: head,
                            agent: agent
                        })];
                case 2:
                    response = _j.sent();
                    return [3, 4];
                case 3:
                    e_2 = _j.sent();
                    console.error(e_2);
                    return [3, 4];
                case 4:
                    if (!response) return [3, 14];
                    _j.label = 5;
                case 5:
                    _j.trys.push([5, 13, , 14]);
                    if (!(typeof process !== "undefined")) return [3, 6];
                    _d = response;
                    return [3, 8];
                case 6: return [4, response.text()];
                case 7:
                    _d = _j.sent();
                    _j.label = 8;
                case 8:
                    html = _d;
                    if (!html) {
                        return [2, renderErrorHtml({ url: url, server: server, noPage: true })];
                    }
                    $html = (0, cheerio_1.load)(html);
                    if (!!baseHref) {
                        $html("head").prepend("<base target=\"_self\" href=\"".concat(typeof baseHref === "string" ? baseHref : url, "\">"));
                    }
                    inlineMutations_2 = [];
                    _loop_1 = function (key) {
                        if (inline[key]) {
                            var attribute_1 = "src";
                            $html(key).attr(attribute_1, function (_, src) {
                                if (src) {
                                    inlineMutations_2.push({ key: key, attribute: attribute_1, src: src + "" });
                                }
                                return src;
                            });
                        }
                    };
                    for (_i = 0, _e = Object.keys(inline); _i < _e.length; _i++) {
                        key = _e[_i];
                        _loop_1(key);
                    }
                    _f = 0, inlineMutations_1 = inlineMutations_2;
                    _j.label = 9;
                case 9:
                    if (!(_f < inlineMutations_1.length)) return [3, 12];
                    com = inlineMutations_1[_f];
                    key = com.key, attribute = com.attribute, src = com.src;
                    element = "".concat(key, "[").concat(attribute, "=\"").concat(src, "\"]");
                    return [4, mutateSource({ key: element, src: src, safeMode: safeMode }, url, $html, headers)];
                case 10:
                    _j.sent();
                    $html(element).removeAttr(attribute);
                    _j.label = 11;
                case 11:
                    _f++;
                    return [3, 9];
                case 12:
                    $html("[src=\"undefined\"]").removeAttr("src");
                    for (_g = 0, _h = Object.keys(cors); _g < _h.length; _g++) {
                        key = _h[_g];
                        if (cors[key]) {
                            $html(key).attr("crossorigin", cors[key]);
                        }
                    }
                    return [2, $html];
                case 13:
                    e_3 = _j.sent();
                    console.error(e_3);
                    return [3, 14];
                case 14: return [2, renderErrorHtml({ url: url, server: server, noPage: true })];
            }
        });
    });
}
function fetchFrame(model) {
    return __awaiter(this, void 0, void 0, function () {
        var $html, e_4;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    _a.trys.push([0, 2, , 3]);
                    return [4, renderHtml(model, typeof process !== "undefined")];
                case 1:
                    $html = _a.sent();
                    return [2, $html === null || $html === void 0 ? void 0 : $html.html()];
                case 2:
                    e_4 = _a.sent();
                    console.error(e_4);
                    return [3, 3];
                case 3: return [2];
            }
        });
    });
}
exports.fetchFrame = fetchFrame;
function configureResourceControl(appConfig) {
    exports.appSourceConfig = appSourceConfig = {
        cors: __assign(__assign({}, config_1.defaultCorsConfig), appConfig === null || appConfig === void 0 ? void 0 : appConfig.cors),
        inline: __assign(__assign({}, config_1.defaultInlineConfig), appConfig === null || appConfig === void 0 ? void 0 : appConfig.inline)
    };
}
exports.configureResourceControl = configureResourceControl;
function createIframe(_req, res, next) {
    var _this = this;
    res.createIframe = function (model) { return __awaiter(_this, void 0, void 0, function () {
        var $html, e_5;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    _a.trys.push([0, 2, , 3]);
                    return [4, renderHtml(model, true)];
                case 1:
                    $html = _a.sent();
                    if ($html) {
                        res.status(200).send($html.html());
                    }
                    return [3, 3];
                case 2:
                    e_5 = _a.sent();
                    console.error(e_5);
                    return [3, 3];
                case 3: return [2];
            }
        });
    }); };
    next();
}
exports["default"] = createIframe;
